File: leetcode/101-200/101-symmetric-tree/bench/is_symmetric.py
"""Benchmark workload for LeetCode #101 — symmetric tree, Python mirror.
Smaller K (pure-Python is slow); timed separately, not cross-checked."""

from __future__ import annotations

import sys

sys.setrecursionlimit(10000)


class Node:
    __slots__ = ("val", "left", "right")

    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def insert(root, v):
    if root is None:
        return Node(v)
    if v < root.val:
        root.left = insert(root.left, v)
    else:
        root.right = insert(root.right, v)
    return root


def mirror(n):
    if n is None:
        return None
    m = Node(n.val)
    m.left = mirror(n.right)
    m.right = mirror(n.left)
    return m


def copy_tree(n):
    if n is None:
        return None
    m = Node(n.val)
    m.left = copy_tree(n.left)
    m.right = copy_tree(n.right)
    return m


def is_mirror(a, b):
    if a is None:
        return b is None
    if b is None:
        return False
    return a.val == b.val and is_mirror(a.left, b.right) and is_mirror(a.right, b.left)


def is_symmetric(root):
    if root is None:
        return True
    return is_mirror(root.left, root.right)


def main():
    base = [8, 4, 12, 2, 6, 10, 14, 1, 3, 5, 7, 9, 11, 13, 15]
    bn = len(base)
    pool = []
    for i in range(8):
        sub = None
        for k in range(bn):
            sub = insert(sub, base[(k + i) % bn])
        root = Node(0)
        root.left = sub
        root.right = mirror(sub) if (i % 2) == 0 else copy_tree(sub)
        pool.append(root)
    acc = 1
    for _ in range(400000):
        idx = acc % 8
        sym = is_symmetric(pool[idx])
        acc = (acc * 131 + (1 if sym else 0) + 1) % 1000000007
    print(acc)


if __name__ == "__main__":
    main()

File: leetcode/101-200/101-symmetric-tree/bench/is_symmetric.rs
// Benchmark workload for LeetCode #101 — symmetric tree, Rust mirror.
// Box<Node> owned tree + &-borrow traversal (is_symmetric is read-only).
const MOD: i64 = 1_000_000_007;

struct Node {
    val: i64,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn insert(root: Option<Box<Node>>, v: i64) -> Option<Box<Node>> {
    match root {
        None => Some(Box::new(Node { val: v, left: None, right: None })),
        Some(mut n) => {
            if v < n.val {
                n.left = insert(n.left.take(), v);
            } else {
                n.right = insert(n.right.take(), v);
            }
            Some(n)
        }
    }
}
fn mirror(n: &Option<Box<Node>>) -> Option<Box<Node>> {
    n.as_ref().map(|node| Box::new(Node {
        val: node.val,
        left: mirror(&node.right),
        right: mirror(&node.left),
    }))
}
fn copy_tree(n: &Option<Box<Node>>) -> Option<Box<Node>> {
    n.as_ref().map(|node| Box::new(Node {
        val: node.val,
        left: copy_tree(&node.left),
        right: copy_tree(&node.right),
    }))
}
fn is_mirror(a: &Option<Box<Node>>, b: &Option<Box<Node>>) -> bool {
    match (a, b) {
        (None, None) => true,
        (Some(an), Some(bn)) => {
            an.val == bn.val && is_mirror(&an.left, &bn.right) && is_mirror(&an.right, &bn.left)
        }
        _ => false,
    }
}
fn is_symmetric(root: &Option<Box<Node>>) -> bool {
    match root {
        None => true,
        Some(n) => is_mirror(&n.left, &n.right),
    }
}
fn main() {
    let base: [i64; 15] = [8, 4, 12, 2, 6, 10, 14, 1, 3, 5, 7, 9, 11, 13, 15];
    let bn = base.len() as i64;
    let mut pool: Vec<Option<Box<Node>>> = Vec::new();
    for i in 0..8i64 {
        let mut sub: Option<Box<Node>> = None;
        for k in 0..bn {
            sub = insert(sub, base[((k + i) % bn) as usize]);
        }
        let other = if (i % 2) == 0 { mirror(&sub) } else { copy_tree(&sub) };
        let root = Some(Box::new(Node { val: 0, left: sub, right: other }));
        pool.push(root);
    }
    let mut acc: i64 = 1;
    for _ in 0..8_000_000 {
        let idx = (acc % 8) as usize;
        let sym = is_symmetric(&pool[idx]);
        acc = (acc * 131 + if sym { 1 } else { 0 } + 1) % MOD;
    }
    println!("{}", acc);
}
